feat(settings): persist settings to localStorage

MichaelBuhler opened the pull request:
Move the settings in the SettingsModal out of the `useGlobalState` `zustand` store and into a new `useSettings` `zustand` store.

Subscribe the `useGlobalState` store to get reactive update on changes made in the SettingsModal.

Use the standard `zustand` middleware to persist the settings from the SettingsModal into the browser's `localStorage`.

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Introduced a dedicated settings management system, allowing for centralized and persistent handling of key application settings.

* **Refactor**
  * Updated various components and hooks to use the new settings management system for retrieving and updating configuration values.
  * Streamlined global state management to separate settings from other application state.
  * Enhanced sidebar state persistence by integrating it into the centralized settings store.

* **Chores**
  * Improved maintainability by reorganizing state management logic and cleaning up exports.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

MichaelBuhler commented:
@kunstmusik The reason for this PR is to persist some settings from the Settings modal into local storage, so that the settings are retained when refreshing the page/navigating back to the portal. The inspiration is that the [Atlas](https://github.com/ar-io/ar-io-grants/issues/37) improvement will have several settings that are worth persisting for reuse.

The meat of this PR is pulling the settings which _are_ configureable in the `SettingModal` popup into their own SettingsState store, so that this group may be serialized as JSON and persisted to local storage. Because of the way that the GlobalState store is implemented and used throughout the app, it needs to subscribe to changes in the settings and instantiate some new objects when the settings change.

The entire GlobalStateInterface state object is not serializable due to recursive references in some objects instantiated from classes and having prototype methods.

kunstmusik commented on .github/workflows/pr.yml:
```diff
@@ -37,6 +37,7 @@ jobs:
 
           VITE_GITHUB_HASH: ${{ github.sha }}
           VITE_AO_CU_URL: ${{ vars.VITE_AO_CU_URL }}
+          VITE_ARIO_PROCESS_ID: 'GaQrvEMKBpkjofgnBi_B3IgIDmY_XYelVLB6GcRGrHc'
```
Think we don't need this in the PR action. I added a suggestion in settings.ts to restore code to constants.ts. Would be alright with switching to devnet by default there. (Will add another comment)

kunstmusik commented:
@MichaelBuhler I left some comments around ARIO_PROCESS_ID in constants.ts. This should be used for initializing settings. (We use this as part of our production build, overriding with VITE_ARIO_PROCESS_ID env var.) I think this can be changed to target testnet by default now.

MichaelBuhler commented on src/store/globalState.ts:
```diff
@@ -0,0 +1,199 @@
+    );
+  }
+
+  setTheme = (theme: ThemeType) => {
+    this.set({ theme });
+    // disabling as this should not be done in the store
+    // applyThemePreference(theme);
+  };
```
At this time, there is no code that actually switches the theme. I will leave this code as-is (as previously implemented).

MichaelBuhler commented on src/store/globalState.ts:
```diff
@@ -0,0 +1,199 @@
+     * If/When: The user changes the AO CU URL or the AR.IO Process in the SettingsModal,
+     * Then: Reinstantiate the AR.IO read and write SDKs.
+     */
+    useSettings.subscribe(
+      ({ aoCuUrl, arioProcessId }) => ({ aoCuUrl, arioProcessId }),
+      ({ aoCuUrl, arioProcessId }) => {
+        const { contractSigner } = get();
+        const arIOReadSDK = makeArIOReadSDK({ aoCuUrl, arioProcessId });
+        const arIOWriteableSDK = makeArIOWriteSDK({
+          aoCuUrl,
+          arioProcessId,
+          contractSigner,
+        });
+        set({ arIOReadSDK, arIOWriteableSDK });
+      },
+    );
+    /* Subscribe to changes in the Settings store.
+     * If/When: The user changes the AR.IO Process in the SettingsModal,
+     * Then: Close and re-open the Dexie (IndexedDB) database.
+     */
+    useSettings.subscribe(
+      ({ arioProcessId }) => arioProcessId,
+      (arioProcessId) => {
+        const currentEpoch = undefined;
+        let { networkPortalDB } = get();
+        networkPortalDB.close();
+        networkPortalDB = createDb(arioProcessId);
+        set({ currentEpoch, networkPortalDB });
+      },
+    );
```
It's true that unsubscribing is good practice. However, `GlobalStateActionBase` is effectively instantiated as a singleton, so we don't need this.

MichaelBuhler commented on src/constants.ts:
```diff
@@ -17,9 +15,6 @@ export const WRITE_OPTIONS = {
   ],
 };
 export const ARIO_DOCS_URL = 'https://docs.ar.io';
-export const ARIO_PROCESS_ID = new ArweaveTransactionID(
```
Done.

kunstmusik commented:
@MichaelBuhler the changes look good, thanks! Also, good note about localStorage and the sidebar. I think it'd make sense to move that to zustand with persist. The one thought I had is whether it'd be good to have a separate item in storage for this vs. settings (i.e., something like { name: "window-state"}). I imagine there are other things that will need to be persisted that are not settings per-se, such as table column filters (eventually).  So my preference is that it would have it's own item-name/hook/setter, but if you want to add to settings for now it can always get refactored later and that's fine to get this work merged in. (Or if you'd like to do that in a separate PR, can accept this one now.)

MichaelBuhler commented:
Fixed a bug with unnecessary state updates (and therefore duplicate network requests). And moved `sidebarOpen` into the Settings store for now; happy to refactor later.

kunstmusik reviewed (approved):
Looks good, thanks @MichaelBuhler for the contribution!
